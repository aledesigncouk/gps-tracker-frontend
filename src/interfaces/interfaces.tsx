export type Coordinate = [number, number];

export interface Geometry {
  type: "LineString";
  coordinates: Coordinate[];
}

export interface Properties {
  year: number;
}

export interface Track {
  type: "Feature";
  properties: Properties;
  geometry: Geometry;
}