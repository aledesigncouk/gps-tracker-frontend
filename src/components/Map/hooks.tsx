import { useState, useEffect } from "react";
import { formatDate, getTrackByRange, setRangeByYear } from "@utils/utils";
import { ControlSwitchEnum } from "@enums/enums";
import { validateTrackData } from "@utils/utils";

type Track = {
  geometry: {
    coordinates: [number, number][];
  };
};

export const useFetchTrack = (
  controlSwitch: string,
  startDate: Date | null,
  endDate: Date | null,
  selectedYear: string | null,
  runFetchData: boolean,
  setRunFetchData: (value: boolean) => void
) => {
  const [track, setTrack] = useState<Track | null>(null);
  const [error, setError] = useState<string | null>(null);


  useEffect(() => {
    const fetchData = async () => {
      try {
        const { start, end } =
          controlSwitch === ControlSwitchEnum.RANGE
            ? { start: formatDate(startDate), end: formatDate(endDate) }
            : (() => {
                const { startDate, endDate } = setRangeByYear(selectedYear);
                return { start: startDate, end: endDate };
              })();

        const result = await getTrackByRange(start, end);
        
        if (validateTrackData(result)) {
          setTrack(result);
        }
        setError(null);
      } catch (error) {
        setError(error.message);
        setTrack(null);
      } finally {
        setRunFetchData(false);
      }
    };

    if (runFetchData || selectedYear) {
      fetchData();
    }
  }, [selectedYear, runFetchData]);

  return { track, error };
};
