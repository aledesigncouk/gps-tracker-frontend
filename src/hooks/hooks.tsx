import { useState, useEffect } from "react";
import { formatDate, getTrackByRange, setRangeByYear } from "@utils/utils";

import { validateTrackData, Track } from "@utils/utils";

export const useFetchTrack = (
  startDate: Date | null,
  endDate: Date | null,
  selectedYear: string | null,
  runFetchData: boolean,
  setRunFetchData: (value: boolean) => void
) => {
  const [track, setTrack] = useState<Track | null>(null);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    const fetchData = async () => {
      try {
        const result = await getTrackByRange(formatDate(startDate), formatDate(endDate));

        if (validateTrackData(result)) {
          setTrack(result);
        }
        setError(null);
      } catch (error) {
        setError(error.message);
        setTrack(null);
      } finally {
        setRunFetchData(false);
      }
    };

    if (runFetchData || selectedYear) {
      fetchData();
    }
  }, [selectedYear, runFetchData]);

  return { track, error };
};
