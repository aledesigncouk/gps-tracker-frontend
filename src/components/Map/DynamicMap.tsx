import { JSX, useEffect, useState } from "react";
import Leaflet from "leaflet";
import * as ReactLeaflet from "react-leaflet";
import { Polyline } from "react-leaflet";
import { useFetchTrack } from "@hooks/hooks";

import "leaflet/dist/leaflet.css";
import styles from "@styles/components/Map.module.scss";
import { Track } from "@/interfaces/interfaces";
import Button from 'react-bootstrap/Button';
import Modal from 'react-bootstrap/Modal';

const { MapContainer } = ReactLeaflet;
const red = { color: "red" };

type MapProps = {
  children: (React: typeof ReactLeaflet, L: typeof Leaflet) => JSX.Element;
  className?: string;
  width?: string;
  height?: string;
  track: Track;
};

const Map: React.FC<MapProps> = ({
  children,
  className,
  width,
  height,
  ...rest
}) => {
  let mapClassName = styles.map;

  if (className) {
    mapClassName = `${mapClassName} ${className}`;
  }

  const { track, error } = useFetchTrack();
  const [modalShow, setModalShow] = useState(false);

  useEffect(() => {
    (async function init() {
      delete Leaflet.Icon.Default.prototype._getIconUrl;
      Leaflet.Icon.Default.mergeOptions({
        iconRetinaUrl: "leaflet/images/marker-icon-2x.png",
        iconUrl: "leaflet/images/marker-icon.png",
        shadowUrl: "leaflet/images/marker-shadow.png",
      });
    })();
  }, []);

  useEffect(() => {
    if (track?.geometry.coordinates.length === 0) {
      setModalShow(true);
    }
  }, [track]);

  return (
    <>
      <MapContainer className={mapClassName} {...rest}>
        {children(ReactLeaflet, Leaflet)}

        {track?.geometry.coordinates.length > 0 &&
          <Polyline
            positions={track?.geometry?.coordinates}
            pathOptions={red}
          />}
      </MapContainer>

      {track?.geometry.coordinates.length === 0 &&
        <Modal
          show={modalShow}
          onHide={() => setModalShow(false)}
          backdrop="static"
          centered
          style={{ zIndex: 2000 }}
          restoreFocus={false}
        >
          <Modal.Header closeButton>
            <Modal.Title>Warning</Modal.Title>
          </Modal.Header>
          <Modal.Body>
            No coordinates found for this track. Please change the date range.
          </Modal.Body>
          <Modal.Footer>
            <Button variant="primary" onClick={() => setModalShow(false)}>
              Close
            </Button>
          </Modal.Footer>
        </Modal>}
    </>
  );
};

export default Map;
