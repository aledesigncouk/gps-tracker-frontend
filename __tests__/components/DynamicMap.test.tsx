import React from "react";
import { render, waitFor, fireEvent } from "@testing-library/react";
import Map from "@components/Map/DynamicMap";
import { useFetchTrack } from "@hooks/hooks";
import { Track } from "@/interfaces/interfaces";

jest.mock("@hooks/hooks", () => ({
    useFetchTrack: jest.fn(),
}));

jest.mock("react-leaflet", () => ({
    MapContainer: jest.fn(({ children }) =>
        <div data-testid="map-container">
            {children}
        </div>),

    Polyline: jest.fn(({ positions }) =>
        <div data-testid="polyline">
            {JSON.stringify(positions)}
        </div>),
}));

describe("DynamicMap component", () => {
    it("render the data and match the snapshot", async () => {
        const track: Track = {
            type: "Feature",
            properties: {
                year: 2022,
            },
            geometry: {
                type: "LineString",
                coordinates: [
                    [52.6609183, -2.4820783],
                    [52.6609183, -2.4820783],
                    [52.6610983, -2.4823683],
                    [52.6612783, -2.4826583],
                    [52.6614583, -2.4829483],
                    [52.6616383, -2.4832383],
                ],
            },
        };
        useFetchTrack.mockReturnValue({ track, error: null });

        const { asFragment, getByTestId } = render(
            <Map
                track={track}
                children={() => (
                    <div data-testid="mock-children">Mocked Children</div>
                )}
            />
        );

        await waitFor(() =>
            expect(getByTestId("map-container")).toBeInTheDocument()
        );

        await waitFor(() =>
            expect(getByTestId("polyline").textContent).toContain(
                JSON.stringify(track.geometry.coordinates)
            )
        );

        expect(asFragment()).toMatchSnapshot()
    });

    it("displays a modal if the track has no coordinates", async () => {
        const track: Track = {
            type: "Feature",
            properties: {
                year: 2022,
            },
            geometry: {
                type: "LineString",
                coordinates: [],
            },
        };

        useFetchTrack.mockReturnValue({ track, error: null });

        const { getByText } = render(
            <Map
                track={track}
                children={() => (
                    <div data-testid="mock-children">Mocked Children</div>
                )}
            />
        );

        await waitFor(() =>
            expect(getByText("Warning")).toBeInTheDocument()
        );
        expect(
            getByText("No coordinates found for this track. Please change the date range.")
        ).toBeInTheDocument();

        expect(getByText("Close")).toBeInTheDocument();
    });

});
