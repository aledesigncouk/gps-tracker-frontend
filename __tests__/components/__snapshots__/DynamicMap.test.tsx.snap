// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicMap component render the data and match the snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="map-container"
  >
    <div
      data-testid="mock-children"
    >
      Mocked Children
    </div>
    <div
      data-testid="polyline"
    >
      [[52.6609183,-2.4820783],[52.6609183,-2.4820783],[52.6610983,-2.4823683],[52.6612783,-2.4826583],[52.6614583,-2.4829483],[52.6616383,-2.4832383]]
    </div>
  </div>
</DocumentFragment>
`;
